opal/mca/common/ofi/common_ofi.c

```diff
@@ -73,7 +103,95 @@ int opal_common_ofi_register_mca_variables(const mca_base_component_t *component
                                       component->mca_type_name,
                                       component->mca_component_name,
                                       "provider_exclude", 0);
+        mca_base_var_register_synonym(verbose_index, component->mca_project_name,
+                                      component->mca_type_name,
+                                      component->mca_component_name,
+                                      "verbose", 0);
     }
 
     return OPAL_SUCCESS;
 }
+
+OPAL_DECLSPEC void opal_common_ofi_mca_register(void)
+{
+    opal_common_ofi.registered++;
+    if (opal_common_ofi.registered > 1) {
+        /* process once */
+        return;
+    }
+
+    opal_common_ofi.output = opal_output_open(NULL);
+    opal_output_set_verbosity(opal_common_ofi.output, opal_common_ofi.verbose);
+}
+
+OPAL_DECLSPEC void opal_common_ofi_mca_deregister(void)
+{
+    /* unregister only on last deregister */
+    opal_common_ofi.registered--;
+    assert(opal_common_ofi.registered >= 0);
+    if (opal_common_ofi.registered) {
+        return;
+    }
+    opal_output_close(opal_common_ofi.output);
+}
+
+/*
+ * Select a provider according to the following rules:
+ * - if user has specified an include list, return first provider that's in the include list
+ * - if use has specified an exclude list, return first provider that's not in the exclude list
+ * - in either case Set the next field in the returned provider fi_info to NULL
+ */
+
+OPAL_DECLSPEC struct fi_info*
+opal_common_ofi_select_ofi_provider(struct fi_info *providers, char *framework_name)
+{
+    struct fi_info *prov = providers;
+    char **include_list = NULL;
+    char **exclude_list = NULL;
+
+    if (NULL == prov) {
+        return NULL;
+    }
+
+    if (NULL != *opal_common_ofi.prov_include) {
+        include_list = opal_argv_split(*opal_common_ofi.prov_include, ',');
+    } else if (NULL != *opal_common_ofi.prov_exclude) {
+        exclude_list = opal_argv_split(*opal_common_ofi.prov_exclude, ',');
+    }
+
+    if (NULL != include_list) {
+        while ((NULL != prov) &&
```

Comment:
Maybe this is existing practice, but it seems that by traversing the providers and checking against the include list we're ignoring any preference the user may have intended. In other words, if I specified an include list of psm2,openib (as an example) and openib happened to appear in the provider list ahead of psm2 we would choose openib when psm2 would have been preferred. 

Again, this might be how it works now, but is that the desirable behavior?

Reply:
We may want to come up with something more complex later but let's stick with current method for this PR.  Anyway I'm thinking of not including the select provider logic change as it looks like there's something like that already in the works - see PR #7641

Reply:
I agree with hprritcha, if we want to do that later, we can. I don't think we specify any ordering priority for include lists, so that is a user assumption right now. We can change that, but I think it'd be more appropriate in a subsequent patch.